internal/pkg/initworkload/workload.go

```diff
@@ -0,0 +1,325 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package initworkload
+
+import (
+	"encoding"
+	"errors"
+	"fmt"
+	"path/filepath"
+
+	"github.com/aws/copilot-cli/internal/pkg/config"
+	"github.com/aws/copilot-cli/internal/pkg/manifest"
+	"github.com/aws/copilot-cli/internal/pkg/term/color"
+	"github.com/aws/copilot-cli/internal/pkg/term/log"
+	"github.com/aws/copilot-cli/internal/pkg/workspace"
+)
+
+const (
+	fmtAddWlToAppStart    = "Creating ECR repositories for %s %s."
+	fmtAddWlToAppFailed   = "Failed to create ECR repositories for %s %s.\n"
+	fmtAddWlToAppComplete = "Created ECR repositories for %s %s.\n"
+)
+
+const (
+	jobWlType = "job"
+	svcWlType = "service"
+)
+
+// Store represents the methods needed to add workloads to the SSM parameter store.
+type Store interface {
+	GetApplication(appName string) (*config.Application, error)
+	CreateService(service *config.Workload) error
+	CreateJob(job *config.Workload) error
+	ListServices(appName string) ([]*config.Workload, error)
+	ListJobs(appName string) ([]*config.Workload, error)
+}
+
+// WorkloadAdder contains the methods needed to add jobs and services to an existing application.
+type WorkloadAdder interface {
+	AddJobToApp(app *config.Application, jobName string) error
+	AddServiceToApp(app *config.Application, serviceName string) error
+}
+
+// Workspace contains the methods needed to manipulate a Copilot workspace.
+type Workspace interface {
+	CopilotDirPath() (string, error)
+	WriteJobManifest(marshaler encoding.BinaryMarshaler, jobName string) (string, error)
+	WriteServiceManifest(marshaler encoding.BinaryMarshaler, serviceName string) (string, error)
+}
+
+// Prog contains the methods needed to render multi-stage operations.
+type Prog interface {
+	Start(label string)
+	Stop(label string)
+}
+
+// WorkloadProps contains the information needed to represent a Workload (job or service)
+type WorkloadProps struct {
+	App            string
+	Type           string
+	Name           string
+	DockerfilePath string
+	Image          string
+
+	Schedule string
+	Timeout  string
+	Retries  int
+
+	Port        uint16
+	HealthCheck *manifest.ContainerHealthCheck
+}
+
+func (p *WorkloadProps) isValidWorkload() bool {
```

Comment:
These validations seem to be overkills to me as init pkg is not used externally and all the inputs are defined and used by us. I think we've already done plenty of validations in the cli pkg and it might be the cli pkg's responsibility to make sure it is passing in the correct inputs.

Another suggestion is we should probably only include workload-wise properties within `WorkloadProps` such as `App`, `Type`, and `name`. And then we could have `ServiceProps` and `JobProps`. With this way we can also make `Service` to take in `ServiceProps` as opposed to `WorkloadProps`. What do you think?

Reply:
hmm, I'd prefer to be able to avoid code duplication in `initWorkload()`--i'm not sure how I'd split the `Props` struct up while still preserving the single flow through that function.

Reply:
Pretty much the same. It won't introduce code duplication i think.

```go
type workloadProps struct {
	App            string
	Type           string
	Name           string
	DockerfilePath string
	Image          string

	Schedule string
	Timeout  string
	Retries  int

	Port        uint16
	HealthCheck *manifest.ContainerHealthCheck
}

type ServiceProps struct {
        *workloadProps
	Port        uint16
	HealthCheck *manifest.ContainerHealthCheck
}

type JobProps struct {
        *workloadProps
	Schedule string
	Timeout  string
	Retries  int
}

// Service writes the service manifest, creates an ECR repository, and adds the service to SSM.
func (s *ServiceInitializer) Service(i *ServiceProps) (string, error) {
	return s.initWorkload(i)
}
```
In this way we can then split `WorkloadInitializer` so that we can have better abstraction level and those createManifest, writeManifest, addWlToApp, addWlToStore could be methods of `WorkloadInitializer` or `ServiceInitializer` or `JobInitializer`, which also enhance the code readability.